crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -1590,9 +1590,16 @@ impl<'db> SpecializationBuilder<'db> {
                             argument: ty,
                         });
                     }
-                    _ => {
-                        self.add_type_mapping(bound_typevar, ty, polarity, f);
-                    }
+                    _ => self.add_type_mapping(bound_typevar, ty, polarity, f),
+                }
+            }
+
+            (Type::SubclassOf(subclass_of), ty) | (ty, Type::SubclassOf(subclass_of))
+                if subclass_of.is_type_var() =>
+            {
+                let formal_instance = Type::TypeVar(subclass_of.into_type_var().unwrap());
+                if let Some(actual_instance) = ty.to_instance(self.db) {
+                    return self.infer_map_impl(formal_instance, actual_instance, polarity, f);
                 }
             }
```

Comment:
This is pretty trivial so I don't think it will overlap much with the constraint solver work, but still flagging it (cc @dcreager).